File: Clone Graph.java
/**
 * Definition for undirected graph.
 * class UndirectedGraphNode {
 *     int label;
 *     ArrayList<UndirectedGraphNode> neighbors;
 *     UndirectedGraphNode(int x) { label = x; neighbors = new ArrayList<UndirectedGraphNode>(); }
 * };
 */
public class Solution {
    /**
     * @param node: A undirected graph node
     * @return: A undirected graph node
     */
    public UndirectedGraphNode cloneGraph(UndirectedGraphNode node) {
        // write your code here
        if (node == null) {
            return node;
        }
        Queue<UndirectedGraphNode> nodes = new LinkedList<UndirectedGraphNode>();
        Map<UndirectedGraphNode, UndirectedGraphNode> map = new HashMap<UndirectedGraphNode, UndirectedGraphNode>();
        nodes.offer(node);
        map.put(node, new UndirectedGraphNode(node.label));
        while (!nodes.isEmpty()) {
            //int size = nodes.size();
            //for (int i = 0; i < size; i++) {
                UndirectedGraphNode oldNode = nodes.poll();
                UndirectedGraphNode newNode = map.get(oldNode);
                for (int j = 0; j < oldNode.neighbors.size(); j++) {
                    UndirectedGraphNode neighbor = oldNode.neighbors.get(j);
                    if (!map.containsKey(neighbor)) {
                        map.put(neighbor, new UndirectedGraphNode(neighbor.label));
                        nodes.offer(neighbor);
                    }
                    newNode.neighbors.add(map.get(neighbor));
                }
           // }
        }
        return map.get(node);
    }
}

File: Clone Graph.cpp
/**
 * Definition for undirected graph.
 * struct UndirectedGraphNode {
 *     int label;
 *     vector<UndirectedGraphNode *> neighbors;
 *     UndirectedGraphNode(int x) : label(x) {};
 * };
 */
class Solution {
public:
    /**
     * @param node: A undirected graph node
     * @return: A undirected graph node
     */
    UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
        // write your code here
        if(node == NULL) return NULL;
         unordered_map<UndirectedGraphNode *, UndirectedGraphNode *> nodeMap;
         queue<UndirectedGraphNode *> visit;
         visit.push(node);
         UndirectedGraphNode * nodeCopy = new UndirectedGraphNode(node->label);
         nodeMap[node] = nodeCopy;
         while (visit.size()>0)
         {
             UndirectedGraphNode * cur = visit.front();
             visit.pop();
             for (int i = 0; i< cur->neighbors.size(); ++i)
             {
                 UndirectedGraphNode * neighb = cur->neighbors[i];
                 if (nodeMap.find(neighb) == nodeMap.end())
                 {
                     // no copy of neighbor node yet. create one and associate with the copy of cur
                     UndirectedGraphNode* neighbCopy = new UndirectedGraphNode(neighb->label);
                     nodeMap[neighb] = neighbCopy;
                     visit.push(neighb);
                 }
                 nodeMap[cur]->neighbors.push_back(nodeMap[neighb]);
             }
         }
     
         return nodeCopy;
    }
};